wagtail/blocks/stream_block.py

```diff
@@ -433,6 +441,17 @@ def deconstruct(self):
         kwargs = self._constructor_kwargs
         return (path, args, kwargs)
 
+    def deconstruct_with_lookup(self, lookup):
```

Comment:
Can StreamBlock's `deconstruct` method be removed with this change? Is it ever called?

With these changes Wagtail now uses BlockDefinitionLookupBuilder to do the deconstruction, which calls `deconstruct_with_lookup`. For most blocks that inherit from Block, that falls back to the block class' `deconstruct`. But blocks like StreamBlock, TypedTableBlock, StructBlock -- basically any block class with a `deconstruct_with_lookup` -- are their regular `deconstruct` methods used anymore?

I suppose they might be useful if there's some code that deconstructs blocks for some other field type besides StreamField, but I can't think of anywhere that might happen. I might be missing something, but it seems like if a block class defines `deconstruct_with_lookup` it no longer needs to define `deconstruct`.

Because Wagtail controls the definition of `Block.deconstruct()` one alternate implementation could be to modify that class' signature to be `Block.deconstruct(lookup=None)` and then use [`accepts_kwarg`](https://github.com/wagtail/wagtail/blob/4e2afc966bc4b37dd425d99c83cfe715563ffcc8/wagtail/coreutils.py#L179) when calling it; this plus a release note could provide a clearer path for third-party packages that implement their own block deconstruction. This feels a little too clever to me, though, versus the clearer distinction between two methods`with_lookup` and without -- I'm just unsure what if means if a class implements both.

Reply:
Currently `deconstruct` is used as part of the definition of `__eq__` on blocks - blocks compare as equal if they have the same signature when deconstructed, and the handful of mutable attributes (such as `name`, `min_num` and `max_num`) are also equal. I don't know if there's actually a valid use for equality testing, though - it was originally introduced to work around a Django <1.9 bug, and then kept in place because it was potentially useful for this exact feature - which turned out not to be the case. (So in fact I should update that code comment...) It's possible that there's some code out there depending on it, so it makes sense to leave it in place.

In any case, `deconstruct` is doing something that's meaningful and well defined outside the context of migrations, namely "return a representation of a method signature that allows copies of this object to be created", and just having it in place as a stepping-stone to understand what `deconstruct_with_lookup` is doing is worthwhile in itself (in much the same way that `to_python` isn't strictly needed now that we have `bulk_to_python`, but `bulk_to_python` is most easily understood in terms of `to_python`).

Reply:
Thanks, that makes sense!